app/src/main/java/org/mozilla/tv/firefox/MainActivity.kt

```diff
@@ -62,19 +62,19 @@ class MainActivity : LocaleAwareAppCompatActivity(), OnUrlEnteredListener, Media
 
         setContentView(R.layout.activity_main)
 
-        val intent = SafeIntent(intent)
-        val intentData = IntentValidator.validateOnCreate(this, intent, savedInstanceState)
+        val intentData = IntentValidator.validateOnCreate(this, intent.toSafeIntent(), savedInstanceState)
 
         val session = getOrCreateSession(intentData)
-
         val screenController = serviceLocator.screenController
         screenController.setUpFragmentsForNewSession(supportFragmentManager, session)
 
-        if (intentData != null) {
-            screenController.showBrowserScreenForUrl(supportFragmentManager, intentData.url)
-        } else {
-            screenController.showBrowserScreenForCurrentSession(supportFragmentManager, session)
-        }
+        serviceLocator.intentLiveData.observe(this, Observer {
+            if (it != null) {
+                screenController.showBrowserScreenForUrl(supportFragmentManager, it.url)
+            } else {
+                screenController.showBrowserScreenForCurrentSession(supportFragmentManager, session)
+            }
+        })
```

Comment:
Interesting, I find having all `screenController` operations in the LiveData observer makes more logical sense

Reply:
I can see both sides on it, but these are the benefits I see for breaking it up.

- As much as I like observables, they're harder to reason about than imperative code.  The more we can get this to read top to bottom, the better
- Handling new startups and handling intents do seem distinct to me.  There's a reason we have to handle intents specifically to avoid lifecycle weirdness, but if the other case just works, why complicate it?
- IMO, it makes the observer easier to understand.  Any time an intent goes into it, it calls `ScreenController#showBrowserScreenForUrl`.  Removing the branching there simplifies potentially confusing code
- We're calling our `LiveData` 'intentData', but as is that doesn't really describe its role.  'intentOrStartData' seems more accurate, which I would consider a code smell.  It seems like it's doing two things in a vaguely odd way.

That said, I don't think it's a huge deal, and I'll be comfortable with whatever you decide upon.

BTW, I edited the code above (added the last line).  Meant to include it yesterday, made a mistake.

Reply:
I do see the benefit of what you're saying for the current code base. One major benefit of having one place for all screenController access based on observable intentData is having one bundled place for all screen controller logic - something like a single dispatcher. This means moving forward, anything additional we add (like Push or Fling), we don't have to dig around the code base looking for places or logics to add. And for the lifecycle weirdness concerns, that's the reason why we use LiveData :)